packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts

```diff
@@ -3,75 +3,124 @@
   return value === undefined;
 };
 
-const group = (existingRow: Row, field: string, value: FieldValue) => {
-  existingRow[field] = value;
+const filterUndefinedAndNull = (values: FieldValue[]): NonNullable<FieldValue[]> => {
+  const filteredValues: NonNullable<FieldValue[]> = [];
+  values.forEach(value => {
+    if (value !== undefined && value !== null) {
+      filteredValues.push(value);
+    }
+  });
+  return filteredValues;
+};
+
+const checkIsNum = (values: FieldValue[]): number[] => {
+  const filteredUndefinedAndNullValues = filterUndefinedAndNull(values);
+  const checkedValues: number[] = [];
+  filteredUndefinedAndNullValues.forEach(value => {
+    if (typeof value !== 'number') {
+      throw new Error(`Expected number, got '${typeof value}'.`);
+    }
+    checkedValues.push(value);
+  });
+  return checkedValues;
 };
 
-const sum = (existingRow: Row, field: string, value: FieldValue) => {
-  if (typeof value === 'number') {
-    existingRow[field] = ((existingRow[field] as number) || 0) + value;
-  } else {
-    throw new Error(`Expected number, got '${typeof value}'.`);
+const group = (values: FieldValue[]): FieldValue => {
+  if (values.length !== 0) {
+    return values[0];
   }
+  return undefined;
 };
 
-const count = (existingRow: Row, field: string, value: FieldValue) => {
-  existingRow[field] = ((existingRow[field] as number) || 0) + 1;
+const sum = (values: FieldValue[]): number | undefined => {
+  if (values.length !== 0) {
+    const checkedValues: number[] = checkIsNum(values);
+    return checkedValues.reduce((a, b) => {
+      return a + b;
+    });
+  }
+  return undefined;
 };
 
-const max = (existingRow: Row, field: string, value: FieldValue) => {
-  const existingValue: FieldValue = existingRow[field];
-  if (!isUndefined(value)) {
-    if (isUndefined(existingValue)) {
-      existingRow[field] = value;
-    } else if (value > existingValue) {
-      existingRow[field] = value;
-    }
+const avg = (values: FieldValue[]): number | undefined => {
+  const numerator: number | undefined = sum(values);
+  const denominator: number = count(values);
+  if (!isUndefined(numerator) && !isUndefined(denominator) && denominator !== 0) {
+    return numerator / denominator;
   }
+  return undefined;
 };
 
-const min = (existingRow: Row, field: string, value: FieldValue) => {
-  const existingValue: FieldValue = existingRow[field];
-  if (!isUndefined(value)) {
-    if (isUndefined(existingValue)) {
-      existingRow[field] = value;
-    } else if (value < existingValue) {
-      existingRow[field] = value;
-    }
+const count = (values: FieldValue[]): number => {
+  return values.length;
+};
+
+const max = (values: FieldValue[]): FieldValue => {
+  if (values.length !== 0) {
+    let maxValue: FieldValue = values[0];
+    values.forEach(value => {
+      if (!isUndefined(value) && !isUndefined(maxValue) && value > maxValue) maxValue = value;
+    });
+    return maxValue;
```

Comment:
```suggestion
    return Math.max(...values.filter(isNotNull));
```

Reply:
I feel like we want to compare string as well? e.g. for period `20210101`, [link to test case](https://github.com/beyondessential/tupaia/blob/3f3f0152152401d6897e4501e1d24f745a624c20/packages/report-server/src/__tests__/reportBuilder/transform/aggregate.test.ts#L107)

but we can do a Array.sort instead of using retro loop 😁

Reply:
Ah, great point! Tbh I was a bit lazy and had assumed that Math.max/min supported string compare as well, which is wrong!

You are correct, and I like the idea of using `Array.sort`. Couple things to keep in mind is that we need to handle `null`s. A good question could be: should `min` return `null` if it's present or the lowest non-null value that exists..? I honestly am not sure, keen to hear your thoughts. I think for now either way is acceptable, as using `null` in our system is still fairly uncommon...

Reply:
Thanks Rohan, I am on your side that it is uncommon to return `null` as minimum value, I think we should make the logic consistent as below to return `null` as min value, the trade off is we need to filter them before the calculation transform, which shouldn't be too hard. 

One case that did happen is we want to get min from array [1, 2, 3, null] -> 1, and array [null, null, null, null] -> null. One way to do it is to eliminate all null value in report server, and then use the fill emtpy period in aggregator.

Reply:
>One way to do it is to eliminate all null value in report server, and then use the fill emtpy period in aggregator.

I think we will need to expect the report-server to handle `null`s. Going forward, client's wont be able to query the aggregator directly at all, so all data will be coming from the report-server.

>One case that did happen is we want to get min from array [1, 2, 3, null] -> 1, and array [null, null, null, null] -> null. 

That's a good point though, but if it comes down to it, we can always support multiple `min` aggregations, eg. `min` and `minNotNull` so that user can choose which one to use when 👍 

For now, I think I'm slightly leaning towards `min` being `min([1,2,3,null]) => null` as this seems to be the most basic function, but am probably happy either way 😄

Reply:
That's a good point, we can do min() and minNotNull() in the future.